src/cargo/ops/registry/publish.rs

```diff
@@ -797,28 +797,51 @@ fn package_list(pkgs: impl IntoIterator<Item = PackageId>, final_sep: &str) -> S
 
 fn resolve_registry_or_index(
     opts: &PublishOpts<'_>,
-    just_pkgs: &[&Package],
+    pkgs: &[&Package],
 ) -> CargoResult<Option<RegistryOrIndex>> {
-    Ok(match opts.reg_or_index.clone() {
-        Some(r) => {
-            validate_registry(&just_pkgs, Some(&r))?;
-            Some(r)
-        }
-        None => {
-            let reg = super::infer_registry(&just_pkgs)?;
-            validate_registry(&just_pkgs, reg.as_ref())?;
-            if let Some(RegistryOrIndex::Registry(registry)) = &reg {
-                if registry != CRATES_IO_REGISTRY {
-                    // Don't warn for crates.io.
-                    opts.gctx.shell().note(&format!(
-                        "found `{}` as only allowed registry. Publishing to it automatically.",
-                        registry
-                    ))?;
+    let opt_index_or_registry = opts.reg_or_index.clone();
+    let registry_is_specified_by_any_package = pkgs
+        .iter()
+        .any(|pkg| pkg.publish().as_ref().map(|v| v.len()).unwrap_or(0) > 0);
+
+    let res = match (opt_index_or_registry, registry_is_specified_by_any_package) {
```

Comment:
nit: I would have split out into a separate refactor commit
- pulling out `opt_index_or_registry`
- pulling out `res`

so it doesn't distract from the main change